# Information Retreival: 

# L1

## What is information Retrieval?

“Information retrieval is a field concerned with
the structure, analysis, organization, storage,
and retrieval of information."

*Gerald Salton*

## ***What is structure in the context of IR?***

+ In the context of Information Retrieval (IR), “structure” refers to the way information or data is organized to facilitate efficient retrieval and relevance ranking. It plays a critical role in enabling IR systems to understand, index, and search through vast amounts of data.

### Document Structure: 

+ Refers to the way individual documents are organized. For example, books may have chapters, sections, and paragraphs, while web pages may have HTML tags indicating headings, links, or metadata.

### Collection Structure: 

+ A collection structure refers to the organization, composition, and properties of the dataset or document corpus being searched. The collection structure plays a critical role in determining how the retrieval system processes, indexes, and retrieves relevant information.

## Information Retrieval: Organisation

***Examples may include:***

1) Cataloguing/tagging

2) Topic

3) Location

4) Likes

## Information Retrieval: Storage

***Examples may include:***

1) Inverted index

2) Distributed storage

3) Peer-to-peer

4) Privacy, security, censorship

## Important concepts in retrieval are *effectiveness*, *precision*, and *recall*

### *Effectiveness*: 

Effectiveness refers to the overall quality or success of a retrieval system in meeting the user’s information needs. It measures how well the system retrieves relevant documents and avoids irrelevant ones. How do you determine the overall quality or success of a retrieval system? Via **Evaluation Metrics**.

### *Precision*:

Precision = (Number of retrieved docs that are relevant) / (total number of retrieved docs)



### *Recall*:

Recall = (Number of retrieved docs that are relevant) / (Total number of relevant docs) 

### Important Terms: *Crawling, Indexing, Boolean Search, and Ranking*


**Crawling** is the process of automatically discovering, fetching, and storing web pages or documents for indexing. It is often the first step in building an IR system.

**Indexing** involves organizing and storing data retrieved during crawling in a structured format that enables fast and efficient search and retrieval.

**Boolean search** uses logical operators to formulate complex queries for retrieving documents that satisfy specific conditions.

Operators can include: **AND**, **NOT**, **OR**...

**Ranking** refers to the process of ordering retrieved documents by relevance to the user’s query.

***How do these fit together in IR?***


	1.	Crawling: Collects raw data (documents or web pages) for the system.
	2.	Indexing: Organizes the data into a searchable structure.
	3.	Boolean Search: Enables users to craft specific queries for retrieval.
	4.	Ranking: Prioritizes the retrieved results to show the most relevant ones first.
    
## Representation of Documents: *Bag of Words*

The *Bag of Words* (BoW) representation is a simple and widely used method in Information Retrieval (IR) and Natural Language Processing (NLP) for representing text data. It treats a document as a “bag” of its words, disregarding grammar, word order, and syntax but preserving the frequency of each word.

## Example of Bag of Words Representation

## Corpus:
1. "The cat sat on the mat."
2. "The dog barked at the cat."

## Vocabulary:
Unique words: `["The", "cat", "sat", "on", "mat", "dog", "barked", "at"]`

## Bag of Words Representation:

| **Word**  | **Doc 1 (Counts)** | **Doc 2 (Counts)** |
|-----------|--------------------|--------------------|
| The       | 2                  | 2                  |
| cat       | 1                  | 1                  |
| sat       | 1                  | 0                  |
| on        | 1                  | 0                  |
| mat       | 1                  | 0                  |
| dog       | 0                  | 1                  |
| barked    | 0                  | 1                  |
| at        | 0                  | 1                  |

## Vector Representation:
- **Doc 1**: `[2, 1, 1, 1, 1, 0, 0, 0]`
- **Doc 2**: `[2, 1, 0, 0, 0, 1, 1, 1]`


# What is Relevance? 

**Relevance** is the “correspondence” between information needs (queries) and information items (documents, webpages, images etc). But, the exact meaning of relevance depends
on applications:

Is it useful?
Is it topical? 
= interesting
= ?

***Predicting relevance is the central goal of IR***

## What is a Retrieval Model? 

A **retrieval model** in the context of Information Retrieval (IR) refers to a formal framework or set of algorithms used to determine the relevance of documents in response to a user’s query. It defines how documents and queries are represented, how their relevance is computed, and how search results are ranked.

### Key components of a Retrieval Model: 

1.	**Document Representation**: How documents are represented in the system (e.g., as bags of words, vectors, embeddings, etc.).
2.	**Query Representation**: How user queries are modeled, which might be similar to or different from document representations.
3.	**Relevance Estimation**: The method for computing the degree of relevance between a query and a document.
4.	**Ranking Mechanism**: Determines the order in which documents are presented to the user, typically based on their relevance scores.

**Categories of Retrieval Models include**: *Exact-Matching Models*, *Vector Space Models*, *Probabilistic Models*, *Neural Retrieval Models*, *Hybrid Models*.

## How do we define Relevance?

Defining **relevance** in Information Retrieval (IR) involves assessing how well a document satisfies a user’s information need. Several factors contribute to this definition, as relevance is not just about matching query terms. Below are detailed explanations of topicality, novelty, freshness, and authority, key factors in determining relevance:

**Topicality**: Refers to how closely a document’s content matches the topic or subject of the user’s query.

**Novelty**: Refers to the degree to which a document provides new or unique information that has not already been encountered by the user.

**Freshness**: Refers to how recent or up-to-date the information in a document is.

**Authority**: Refers to the credibility, trustworthiness, or expertise of the source or author of a document.

*To summarise*: 

+	**Topicality** ensures that the document matches the query.
+	**Novelty** avoids redundancy and adds unique value.
+	**Freshness** provides time-sensitive information.
+	**Authority** ensures the information is credible and trustworthy.

*An example*: 

For a query like “climate change impacts in 2024”:

+ Topicality ensures documents discuss climate change.
+ Novelty ensures unique insights are retrieved.
+ Freshness prioritizes recent 2024 studies or reports.
+ Authority ensures reliable sources, like government reports or peer-reviewed papers, rank higher.

# Exact Matching in Information Retrieval

## Core Concepts of Exact Matching

1. **Keyword Matching**:
   - Documents are retrieved if they contain the exact keywords specified in the query.
   - **Example**:
     - Query: "machine learning"
     - Retrieved document: Contains the exact phrase "machine learning."

2. **Boolean Matching**:
   - Uses logical operators like AND, OR, and NOT to define conditions for document retrieval.
   - **Examples**:
     - Query: `"machine AND learning"` retrieves documents that contain both "machine" and "learning."
     - Query: `"machine OR learning"` retrieves documents that contain either "machine" or "learning" or both.

3. **Phrase Matching**:
   - Retrieves documents containing a specific sequence of words.
   - **Example**:
     - Query: `"deep neural networks"`
     - Only documents with the exact phrase "deep neural networks" are retrieved.

4. **Wildcard Matching**:
   - Allows partial term matching using special symbols (e.g., `*` or `?`).
   - **Example**:
     - Query: `"comput*"` matches "computer," "computing," "computation," etc.

5. **Field-Specific Matching**:
   - Matches terms in specific fields of a document, such as the title, abstract, or author name.
   - **Example**:
     - Query: `title:"machine learning"`

---

## Strengths of Exact Matching

1. **Simplicity**:
   - Straightforward to implement and understand.
   - Does not require complex algorithms or semantic analysis.

2. **Precision**:
   - High precision when queries are well-constructed, as only documents explicitly matching the query terms are retrieved.

3. **Deterministic**:
   - Provides consistent results since matching is based on strict rules.

4. **Speed**:
   - Efficient for small-scale collections with well-defined indexing.

---

## Limitations of Exact Matching

1. **Vocabulary Mismatch**:
   - Cannot handle cases where users use synonyms or related terms.
   - **Example**:
     - Query: "car" will not retrieve documents containing only "automobile."

2. **Lack of Flexibility**:
   - Overly rigid; slight variations in wording or phrasing can lead to missed documents.
   - **Example**:
     - Query: `"AI research"` may miss documents using "artificial intelligence research."

3. **Context Ignorance**:
   - Ignores the semantic relationships and meaning of words.
   - **Example**:
     - Query: "bass" could retrieve documents about fish or music, with no understanding of context.

4. **No Ranking or Relevance Scoring**:
   - Documents are not ranked based on their relevance but are often returned in arbitrary or predefined order (e.g., by document ID or timestamp).

5. **Challenges with Long Queries**:
   - Long or complex queries may yield no results if all terms are required to match.
   
 
# Ranked Retrieval: 
   
**Ranked retrieval** is a method used in Information Retrieval to organize the results of a search query by their relevance to the user’s query. Unlike exact matching, which often provides unranked or unordered results, ranked retrieval ensures that the most relevant documents appear at the top of the result list.

# Vector Space Model (VSM)

The **Vector Space Model (VSM)** is a mathematical and conceptual framework used in **Information Retrieval (IR)** to represent and compare documents and queries as vectors in a multi-dimensional space. It allows retrieval and ranking of documents based on their relevance to a query.

---

## Core Concepts of the Vector Space Model

1. **Vector Representation**:
   - Each document and query is represented as a vector in a space where each dimension corresponds to a unique term (word) in the vocabulary.
   - The value in each dimension is a weight, often based on the term's importance in the document or query (e.g., TF-IDF).

2. **Similarity Measurement**:
   - The relevance of a document to a query is determined by the similarity between their respective vectors.
   - Similarity is commonly measured using **cosine similarity**, which computes the angle between the two vectors.

3. **Partial Matching**:
   - VSM allows for partial matching, meaning documents can still be considered relevant even if they don't contain all query terms.

---

## Mathematics of the Vector Space Model

### **Document Representation**:
Given a vocabulary of \( n \) terms, each document $ D_i $ is represented as a vector:
$$
\vec{D_i} = [w_{i1}, w_{i2}, ..., w_{in}]
$$
where $ w_{ij} $ is the weight of term $ j $ in document $ i $.

### **Query Representation**:
Similarly, the query $ Q $ is represented as:
$$
\vec{Q} = [q_1, q_2, ..., q_n]
$$
where $ q_j $ is the weight of term $ j $ in the query.

---

### **Term Weighting**:
Common weighting schemes include:

- **Term Frequency (TF)**: Counts the occurrences of a term in a document.
$$
\text{TF}(t, D) = \frac{\text{Frequency of term } t \text{ in } D}{\text{Total terms in } D}
$$

- **Inverse Document Frequency (IDF)**: Measures how unique a term is across the corpus.
$$ 
\text{IDF}(t) = \log\frac{\text{Total number of documents}}{\text{Number of documents containing } t}
$$

- **TF-IDF**: Combines TF and IDF to prioritize terms that are frequent in a document but rare in the corpus.
$$
\text{TF-IDF}(t, D) = \text{TF}(t, D) \times \text{IDF}(t)
$$

---

### **Cosine Similarity**:
Cosine similarity measures the cosine of the angle between the document vector \( \vec{D_i} \) and query vector \( \vec{Q} \):
$$
\text{Cosine Similarity} = \cos(\theta) = \frac{\vec{D_i} \cdot \vec{Q}}{\|\vec{D_i}\| \|\vec{Q}\|}
$$
where:
- $ \vec{D_i} \cdot \vec{Q} $: Dot product of the document and query vectors.
- $ \|\vec{D_i}\| $: Magnitude of the document vector.
- $ \|\vec{Q}\| $: Magnitude of the query vector.

---

## Steps in Using the Vector Space Model

1. **Preprocessing**:
   - Tokenize text into terms.
   - Remove stopwords and apply stemming or lemmatization.

2. **Vectorization**:
   - Construct term-document and term-query matrices.
   - Calculate term weights (e.g., using TF-IDF).

3. **Similarity Computation**:
   - Compute cosine similarity between the query vector and each document vector.

4. **Ranking**:
   - Rank documents in descending order of their similarity scores.

---

## Example

### **Corpus**:
1. **Doc 1**: "The cat sat on the mat."
2. **Doc 2**: "The dog barked at the cat."

### **Vocabulary**:
- Terms: `["the", "cat", "sat", "on", "mat", "dog", "barked", "at"]`

### **Term Frequencies (TF)**:

| **Term**  | **Doc 1 (TF)** | **Doc 2 (TF)** |
|-----------|----------------|----------------|
| the       | 0.33           | 0.33           |
| cat       | 0.17           | 0.17           |
| sat       | 0.17           | 0.00           |
| on        | 0.17           | 0.00           |
| mat       | 0.17           | 0.00           |
| dog       | 0.00           | 0.17           |
| barked    | 0.00           | 0.17           |
| at        | 0.00           | 0.17           |

---

## Advantages of the Vector Space Model

1. **Flexible Matching**:
   - Supports partial matches, increasing recall.

2. **Quantifiable Similarity**:
   - Provides a numeric relevance score for ranking.

3. **Conceptually Simple**:
   - Intuitive and easy to implement.
   
# Zipf's Law for Term Frequency

**Zipf's Law** is a principle that describes the statistical distribution of word frequencies in natural language text. It states that the frequency of a word is **inversely proportional to its rank** in the frequency distribution. In simpler terms, the most frequent word appears roughly twice as often as the second most frequent word, three times as often as the third most frequent word, and so on.

---

## Formal Definition

If:
- \( f \): frequency of a word,
- \( r \): rank of the word in the frequency distribution (1 for the most frequent word, 2 for the second most frequent, etc.),
- \( C \): a constant of proportionality,

then Zipf's Law is expressed as:
$$
f(r) \propto \frac{1}{r}
$$
or equivalently:
$$
f(r) = \frac{C}{r}
$$

---

## Characteristics of Zipf's Law

1. **Power-Law Distribution**:
   - Zipf's Law is a specific case of a power-law distribution. It implies that a few words are extremely frequent, while most words are rare.
   - The relationship can be plotted on a log-log scale, where the rank and frequency form approximately a straight line.

2. **High-Frequency Words**:
   - A small number of words (e.g., "the," "and," "is") make up a large proportion of the text.

3. **Long Tail**:
   - Most words in a corpus occur very infrequently (e.g., rare or domain-specific terms).

---

## Example

Consider the following text:

- **Text**: "the cat sat on the mat and the dog barked"
- **Word Frequencies**:
  - "the" (4 occurrences)
  - "cat" (1 occurrence)
  - "sat" (1 occurrence)
  - "on" (1 occurrence)
  - "mat" (1 occurrence)
  - "and" (1 occurrence)
  - "dog" (1 occurrence)
  - "barked" (1 occurrence)

### Rank-Frequency Table:

| **Rank (r)** | **Word** | **Frequency (f)** | $$ f \propto \frac{1}{r} $$ |
|--------------|----------|-------------------|----------------------------|
| 1            | the      | 4                 | $$ 4 = \frac{C}{1} $$      |
| 2            | cat      | 1                 | $$ 1 = \frac{C}{2} $$      |
| 3            | sat      | 1                 | $$ 1 = \frac{C}{3} $$      |
| 4            | on       | 1                 | $$ 1 = \frac{C}{4} $$      |
| ...          | ...      | ...               | ...                        |

Here, the most frequent word "the" has 4 occurrences, which is roughly twice as frequent as the second most frequent word, and so on.

---

## Applications of Zipf's Law in Information Retrieval

1. **Stopword Removal**:
   - Words like "the," "and," and "is" (highly frequent) often provide little information and can be removed to reduce noise.

2. **Index Optimization**:
   - Understanding term frequencies can guide the construction of efficient inverted indexes by prioritizing common terms.

3. **TF-IDF Weighting**:
   - Zipf's Law underpins the **Inverse Document Frequency (IDF)** component of TF-IDF, emphasizing terms that are rare across documents.

4. **Vocabulary Analysis**:
   - Helps understand language characteristics and guides the selection of features for NLP tasks.

---

## Limitations of Zipf's Law

1. **Empirical Approximation**:
   - While Zipf's Law holds well for natural language, it is not an exact law and may deviate for specific corpora or domains.

2. **Assumes Independence**:
   - Ignores relationships between words, such as phrases or semantic dependencies.

3. **Long Tail Handling**:
   - Rare terms in the "long tail" may require different processing strategies, such as smoothing or data augmentation.
   
# Inverse Document Frequency (IDF)

**Inverse Document Frequency (IDF)** is a numerical statistic used in **Information Retrieval (IR)** to measure the importance or rarity of a term across a collection of documents. It is a core component of the widely-used **TF-IDF** weighting scheme, which combines **Term Frequency (TF)** with IDF to rank terms based on their relevance.

---

## Core Concept

- Terms that appear frequently across many documents (e.g., "the," "and") are less informative because they are common.
- Terms that appear in fewer documents (e.g., "quantum mechanics," "photosynthesis") are more informative and significant in identifying the content of a document.
- IDF gives higher weights to less common terms and lower weights to more common terms.

---

## Formula

The IDF of a term $t$ is calculated as:
$$
\text{IDF}(t) = \log\frac{N}{1 + \text{DF}(t)}
$$
where:
- $N$: Total number of documents in the collection.
- $\text{DF}(t)$: Number of documents containing the term $t$ (document frequency).
- $1 + \text{DF}(t)$: Adding 1 to avoid division by zero for terms that do not appear in any document.

---

## Key Points

1. **High IDF**:
   - Terms that appear in very few documents have a high IDF, indicating their importance or uniqueness.
   - Example: If "quantum mechanics" appears in 2 out of 1,000 documents, its IDF is high.

2. **Low IDF**:
   - Common terms (like "the," "and," "is") have a low IDF, reflecting their ubiquity and lack of discriminative power.
   - Example: If "the" appears in all 1,000 documents, its IDF is close to zero.

3. **Logarithmic Scaling**:
   - The logarithm dampens the effect of very high document frequencies to avoid over-penalizing common terms.

---

## Example Calculation

### Given:
- Total documents ($N$): 10,000
- Document Frequency of "machine" ($\text{DF}(\text{machine})$): 1,000
- Document Frequency of "learning" ($\text{DF}(\text{learning})$): 50

### IDF Values:

1. For "machine":
   $$
   \text{IDF}(\text{machine}) = \log\frac{10,000}{1 + 1,000} = \log\frac{10,000}{1,001} \approx 1
   $$

2. For "learning":
   $$
   \text{IDF}(\text{learning}) = \log\frac{10,000}{1 + 50} = \log\frac{10,000}{51} \approx 2.3
   $$

"Learning" has a higher IDF than "machine," indicating that it is a rarer and more informative term.

---

## Why Use IDF?

1. **Improves Relevance**:
   - Helps prioritize terms that are more likely to differentiate between documents.

2. **Reduces Noise**:
   - Common terms (stopwords) have little impact on ranking since their IDF is low.

3. **Balances Term Frequency**:
   - Works with **TF (Term Frequency)** to ensure that rare terms with high TF in a document are emphasized.

---

## Applications of IDF

1. **TF-IDF Weighting**:
   - Used to rank documents in search engines based on relevance.
   - $$
   \text{TF-IDF}(t, D) = \text{TF}(t, D) \times \text{IDF}(t)
   $$

2. **Text Classification**:
   - Features with high TF-IDF values are often used as input for machine learning models.

3. **Topic Modeling**:
   - Helps identify key terms that define a topic in a corpus.

4. **Summarization**:
   - Identifies critical terms in documents for summarization tasks.

---

## Limitations of IDF

1. **Static Nature**:
   - Traditional IDF is calculated based on a fixed corpus, which can be outdated for dynamic or growing datasets.

2. **Fails with Short Documents**:
   - In small or sparse datasets, IDF may not provide meaningful weights.

3. **Ignores Semantics**:
   - IDF focuses on term frequency and does not consider the meaning or context of terms.

4. **Over-Penalization of Common Terms**:
   - Very high document frequencies may result in negligible weights for moderately useful terms.
   
# Cosine Similarity

**Cosine Similarity** is a measure used in **Information Retrieval (IR)** and **Natural Language Processing (NLP)** to determine the similarity between two vectors. It calculates the cosine of the angle between two non-zero vectors in a multi-dimensional space, where smaller angles indicate higher similarity.

---

## Definition

Cosine Similarity measures the orientation (not the magnitude) of two vectors, making it particularly useful when comparing documents where the magnitude of the vectors (e.g., word frequency counts) is less important than their direction.

### Formula
For two vectors $ \vec{A} $ and $ \vec{B} $, cosine similarity is defined as:
$$
\text{Cosine Similarity} = \cos(\theta) = \frac{\vec{A} \cdot \vec{B}}{\|\vec{A}\| \|\vec{B}\|}
$$
Where:
- $ \vec{A} \cdot \vec{B} $: Dot product of vectors $ \vec{A} $ and $ \vec{B} $.
- $ \|\vec{A}\| $: Magnitude (or norm) of vector $ \vec{A} $, calculated as:
  $$
  \|\vec{A}\| = \sqrt{\sum_{i=1}^n A_i^2}
  $$
- $ \|\vec{B}\| $: Magnitude of vector $ \vec{B} $.

---

## Steps to Compute Cosine Similarity

1. **Vector Representation**:
   - Represent the documents or queries as vectors. Each dimension corresponds to a term in the vocabulary, and the vector values are typically weights like term frequency (TF) or TF-IDF.

2. **Compute the Dot Product**:
   - Multiply corresponding components of the vectors and sum the results:
     $$
     \vec{A} \cdot \vec{B} = \sum_{i=1}^n A_i \cdot B_i
     $$

3. **Compute the Magnitudes**:
   - Calculate the Euclidean norm of each vector:
     $$
     \|\vec{A}\| = \sqrt{\sum_{i=1}^n A_i^2}, \quad \|\vec{B}\| = \sqrt{\sum_{i=1}^n B_i^2}
     $$

4. **Calculate Cosine Similarity**:
   - Divide the dot product by the product of the magnitudes.

---

## Example

### Vectors:
Let:
- $ \vec{A} = [1, 2, 3] $
- $ \vec{B} = [4, 5, 6] $

### Step 1: Compute the Dot Product:
$$
\vec{A} \cdot \vec{B} = (1 \cdot 4) + (2 \cdot 5) + (3 \cdot 6) = 4 + 10 + 18 = 32
$$

### Step 2: Compute the Magnitudes:
- $ \|\vec{A}\| $:
  $$
  \|\vec{A}\| = \sqrt{1^2 + 2^2 + 3^2} = \sqrt{1 + 4 + 9} = \sqrt{14}
  $$
- $ \|\vec{B}\| $:
  $$
  \|\vec{B}\| = \sqrt{4^2 + 5^2 + 6^2} = \sqrt{16 + 25 + 36} = \sqrt{77}
  $$

### Step 3: Compute Cosine Similarity:
$$
\text{Cosine Similarity} = \frac{\vec{A} \cdot \vec{B}}{\|\vec{A}\| \|\vec{B}\|} = \frac{32}{\sqrt{14} \cdot \sqrt{77}} \approx \frac{32}{32.86} \approx 0.974
$$

The cosine similarity is approximately 0.974, indicating that the vectors are highly similar.

---

## Properties of Cosine Similarity

1. **Range**:
   - The result lies between -1 and 1:
     - **1**: Vectors are identical in direction.
     - **0**: Vectors are orthogonal (completely dissimilar).
     - **-1**: Vectors point in opposite directions (rare in practice for text data).

2. **Magnitude-Independent**:
   - Cosine similarity focuses on direction, not magnitude, making it robust for comparing documents of different lengths.

3. **Sparsity**:
   - Works well with sparse vectors, as is common in text data.

---

## Applications

1. **Document Similarity**:
   - Measure the similarity between documents for tasks like plagiarism detection or duplicate detection.

2. **Search Ranking**:
   - Rank documents based on their cosine similarity to a query.

3. **Clustering**:
   - Group similar documents in clustering algorithms like k-means.

4. **Recommendation Systems**:
   - Compare user profiles or item attributes to recommend similar items.

---

## Advantages

1. **Robust to Length Differences**:
   - Accounts for differences in document or query lengths, focusing only on term distribution.

2. **Efficient to Compute**:
   - Especially efficient with sparse data and optimized libraries.

3. **Widely Used**:
   - A standard similarity measure in IR, NLP, and machine learning.

---

## Limitations

1. **Ignores Term Relationships**:
   - Considers only the terms explicitly in the vector, ignoring synonyms or semantic relationships.

2. **Sensitive to Vector Sparsity**:
   - Works best when vectors are well-populated; sparse vectors may yield less meaningful results.

3. **Binary or Weighted Values**:
   - Performance depends on how terms are weighted (e.g., binary, TF, or TF-IDF).



















# L2: Prof. Lampos part (First Hour) 

## Text Processing: 

A document unit (book, book chapter, article, paragraph, sentence, tweet, search query, fixed window of terms...) will go through a number of steps: 

+ Parsing 
+ Tokenisation
+ Normalisation
+ Stop Word Removal
+ Lemmatisation
+ Stemming

It must be noted that *Stop Word Removal*, *Lemmatisation*, and *Stemming* are optional and this order is not necessarily rigid. For example, tokenisation may come before parsing. 

### Parsing: 

If the file is not raw text, e.g. JSON, HTML. 

### Tokenisation: 

The task of chopping up a document unit into pieces, called tokens

    Sentence: “They won’t let you fly, but they might let you sing.”
    Token:  [They] [won’t] [let] [you] [fly] [,] [but] [they] [might] [let] [you] [sing] [.]
    
**Tokens** need to be turned to **terms**, i.e. processed tokens that will be
maintained in our vocabulary index.

### Normalisation:

The process of canonicalising tokens so that during indexing matches
occur despite of superficial differences in the character sequences.

### Stop Word Removal: 

Extremely common (very frequent) words that do not add to the meaningof a document unit, but exact defini,on depends on the set of decisions we make (linked to the target task) in order to identify stop words.

*Examples*: “the”, “an”, “to”, “so”, “then”.

*Benefits*: reduces number of features / dimensionality and helps derive
models that can generalise beYer, saves storage / memory space.

*Issues*: might remove some meaning from the text
e.g. “flights to London” — if we remove “to” as a stop word, then we don’t
know whether this text snippet is about flights “to” or “from” London.

‣ Could be determined using a predefined list and/or automa,cally, e.g.
the most frequent terms in very large corpus

‣ Bag-of-words models (each term is considered in isola,on) could benefit
from stop word removal, but modern language models (e.g. BERT or GPT
variants) might not as stop words can add to the seman,c interpreta,on
of text

‣ Should we remove stop words? Depends on the method used and the
target task. Most of the ,mes the downstream task accuracy can be
measured, and we can actually see whether removing stop words helps
or not and how much.

### Lemmatisation: 

Returns the base (*dictionary*) form of a word, which is known as the lemma.
    
    + “organises” or “organising” to “organise”
    + “cars” to “car”
    + “saw” to “see” (if “saw” is a verb)
   
+  Does things “properly”, i.e. requires a complete vocabulary and morphological analysis (needs to know what part of speech is the target word for example), aiming to remove inflectional endings only.

### Stemming: 

Crude heuristic process that uses a stemmer (stemming algorithm) in an
attempt to reduce inflected (or derived) words / tokens to their word
stem (root form) — the stem, i.e. the output of a stemmer is very often not a vocabulary word. 

    + “cars” to “car”
    + “organises” or “organising” to “organis”
    + “story” or “stories” to “stori”
    
# Zipfian Distribution

The **Zipfian Distribution** models the frequency of a word or term $f(r)$ in a dataset, based on its rank $r$ when terms are ordered by frequency. It is given by:

$$
f(r) = \frac{C}{r^s}
$$

Where:
- $f(r)$: Frequency of the term at rank $r$.
- $C$: Normalization constant, ensuring that the probabilities sum to 1 across all ranks.
- $r$: Rank of the term (1 for the most frequent term, 2 for the second most frequent, etc.).
- $s$: Exponent that determines the skew of the distribution (commonly $s \approx 1$ for natural language text).

---

## Normalization Constant ($C$)

To ensure that the frequencies or probabilities sum to 1 across all ranks $r = 1, 2, ..., N$, the normalization constant $C$ is computed as:
$$
C = \frac{1}{\sum_{r=1}^N \frac{1}{r^s}}
$$

Where:
- $N$: Total number of unique terms in the dataset.

---

## Key Points

1. **Zipf's Law as a Special Case**:
   - When $s = 1$, the distribution simplifies to Zipf's Law:
     $$
     f(r) = \frac{C}{r}
     $$

2. **Interpretation**:
   - The term at rank $r = 1$ (the most frequent term) has the highest frequency.
   - As $r$ increases, $f(r)$ decreases exponentially, meaning terms of higher rank appear less frequently.

3. **Applications**:
   - Models word frequency in natural language.
   - Describes power-law distributions in various domains (e.g., city populations, web page visits).




# L2: Probabilistic Retrieval Models 

## Language Models: 

• A language model is a probability distribution over strings of text.

• How likely is a given string in a given language?

• Consider probabilities for the following strings:

    • p1 = P(“language model”)
    • p2 = P(“probability distribution”)
    • p3 = P(“a language model is a probability distribution”)
    
• Every possible string has some probability.
• Depends on the language we are modelling

### What is a Unigram Language Model?

A **unigram language model** assumes that each word in a document (or query) is generated independently of every other word. In this model, the probability of a sequence of words is the product of the probabilities of the individual words:

$$
P(w_1, w_2, \dots, w_n) = P(w_1) \cdot P(w_2) \cdot \dots \cdot P(w_n)
$$

Where $P(w_i)$ is the probability of the word $w_i$ in the model. This assumption of independence simplifies calculations and is what makes it a "unigram" model.

---

### How is it used in Information Retrieval?

In **Information Retrieval (IR)**, unigram language models are often employed to calculate the probability of a query $Q = \{q_1, q_2, \dots, q_m\}$ given a document $D = \{d_1, d_2, \dots, d_n\}$. The core idea is to rank documents based on how likely they are to generate the given query.

This likelihood is calculated using the **language model of the document**. Each document is treated as a "bag of words," and the probability of the query is computed as:

$$
P(Q|D) = \prod_{i=1}^{m} P(q_i|D)
$$

Here:
- $P(q_i|D)$: The probability of the query word $q_i$ occurring in document $D$.

### Bigram Language Model

A **bigram language model** is an extension of the unigram model that takes into account word-to-word dependencies. Instead of assuming that each word is independent of the others, it models the conditional probability of a word based on the previous word in the sequence. This provides a simplistic way to include some context into the language model.

---

### Definition

In a bigram language model, the probability of a sequence of words $w_1, w_2, \dots, w_n$ is calculated as:

$$
P(w_1, w_2, \dots, w_n) = P(w_1) \cdot P(w_2|w_1) \cdot P(w_3|w_2) \cdot \dots \cdot P(w_n|w_{n-1})
$$

Here:
- $P(w_1)$ is the probability of the first word.
- $P(w_i|w_{i-1})$ is the conditional probability of word $w_i$ given the previous word $w_{i-1}$.

---

### How is it Used in Information Retrieval?

In **Information Retrieval (IR)**, a bigram language model can be applied to better capture dependencies between words in a query and a document. This is particularly useful for queries that involve phrases, where the order of words matters.

For a query $Q = \{q_1, q_2, \dots, q_m\}$ and a document $D = \{d_1, d_2, \dots, d_n\}$, the likelihood of the query under the document's language model is calculated as:

$$
P(Q|D) = P(q_1|D) \cdot P(q_2|q_1, D) \cdot P(q_3|q_2, D) \cdot \dots \cdot P(q_m|q_{m-1}, D)
$$

Here:
- $P(q_i|q_{i-1}, D)$ is the conditional probability of $q_i$ given $q_{i-1}$ under the document language model.
- This helps account for word pair (bigram) probabilities in the document and query.

---

### Three main approaches Language MOdels (LM) in IR: 


• **Query likelihood model**: P(Q|M_D)

• What is the probability to generate the given query, given a document language model?

• **Document likelihood model**: P(D|M_Q)

• What is the probability to generate the given document, given a query language model?

• **Divergence model**: D(P(M_D) || P(M_Q))

• How ”close” are 2 statistical models?


### 1. Query Likelihood Model (QLM)

The **Query Likelihood Model** is a probabilistic framework used in **language modeling for information retrieval**. In this approach, the goal is to rank documents based on the probability that a document’s language model generates the given query.

#### Core Idea
Rank documents based on $P(Q|D)$, the probability of the query $Q$ given a document $D$.

#### Mathematical Representation
Using the **unigram model** assumption (independence between query terms), $P(Q|D)$ is computed as:

$$
P(Q|D) = \prod_{i=1}^{m} P(q_i|D)
$$

Where:
- $Q = \{q_1, q_2, \dots, q_m\}$ is the query with $m$ terms.
- $P(q_i|D)$ is the probability of query term $q_i$ under the document’s language model.

#### Estimating $P(q_i|D)$
This is often done using **Maximum Likelihood Estimation (MLE)** with smoothing to handle zero probabilities:
- **Unsmoothed MLE**:
  $$
  P(q_i|D) = \frac{\text{Count}(q_i, D)}{|D|}
  $$
  Where:
  - $\text{Count}(q_i, D)$: Number of times $q_i$ occurs in $D$.
  - $|D|$: Total number of terms in $D$.

- **Smoothing**: Common techniques like **Jelinek-Mercer** or **Dirichlet smoothing** are used to account for unseen terms.

#### Key Applications
- Ranking documents based on their likelihood of generating the query.
- Simple and effective for many IR tasks.

#### Limitations
- Assumes independence between query terms.
- Relies on accurate estimation of $P(q_i|D)$, which can be challenging for short documents or rare terms.

---

### 2. Document Likelihood Model (DLM)

The **Document Likelihood Model** flips the roles of the query and document compared to the Query Likelihood Model. Instead of ranking documents based on the likelihood of generating a query, it ranks queries based on their likelihood of generating a given document.

#### Core Idea
Rank queries $Q$ based on $P(D|Q)$, the probability of the document $D$ given the query $Q$.

#### Mathematical Representation
Using **Bayes’ Theorem**:

$$
P(D|Q) = \frac{P(Q|D) \cdot P(D)}{P(Q)}
$$

- $P(Q|D)$: Likelihood of the query given the document’s language model (as in the QLM).
- $P(D)$: Prior probability of the document (can be uniform or weighted).
- $P(Q)$: Normalizing constant (same for all documents and often ignored in ranking).

#### Key Characteristics
- Puts more emphasis on the document’s prior probability, which can be useful in certain applications like **recommendation systems**.
- Less commonly used in traditional ad hoc retrieval tasks.

---

### 3. Divergence Model

The **Divergence Model** focuses on the **difference** between the query and document language models. Instead of modeling the likelihood directly, it ranks documents based on how similar or dissimilar their language models are to the query’s language model.

#### Core Idea
Rank documents based on the divergence (distance) between the query language model $\theta_Q$ and the document language model $\theta_D$. This is typically done using **Kullback-Leibler Divergence (KL Divergence)**.

#### Mathematical Representation
The **KL Divergence** between $\theta_Q$ and $\theta_D$ is given by:

$$
D_{\text{KL}}(\theta_Q || \theta_D) = \sum_{w \in V} P(w|Q) \log \frac{P(w|Q)}{P(w|D)}
$$

Where:
- $V$: Vocabulary of terms.
- $P(w|Q)$: Probability of term $w$ in the query language model.
- $P(w|D)$: Probability of term $w$ in the document language model.

#### Ranking Criterion
Since smaller divergence means greater similarity, documents are ranked by minimizing the divergence:

$$
\text{Score}(D, Q) = -D_{\text{KL}}(\theta_Q || \theta_D)
$$

#### Estimation of $P(w|Q)$ and $P(w|D)$
- $P(w|Q)$: Often approximated as the relative frequency of $w$ in the query.
- $P(w|D)$: Estimated using **MLE** or smoothing techniques.

#### Advantages
- Incorporates the entire vocabulary $V$, allowing for a more nuanced comparison of models.
- Can capture fine-grained differences between the query and document.

#### Limitations
- Requires good smoothing techniques to handle rare or unseen terms.
- Computationally more expensive than QLM or DLM.

---

### Summary

| Model                    | Ranking Criterion               | Key Concept                                     | Common Use Case                              |
|--------------------------|----------------------------------|------------------------------------------------|---------------------------------------------|
| **Query Likelihood Model** | $P(Q|D)$                       | Rank documents by how likely they are to generate the query. | Ad hoc information retrieval tasks.         |
| **Document Likelihood Model** | $P(D|Q)$                       | Rank documents by their posterior probability given a query. | Recommendation systems, specialized ranking. |
| **Divergence Model**      | $-D_{\text{KL}}(\theta_Q || \theta_D)$ | Rank documents by minimizing divergence between query and document language models. | Complex comparisons of models.      |

### Premise of Smoothing in Language Models

Smoothing in language models addresses the problem of **data sparsity**. The core idea is to adjust probability estimates so that rare or unseen events (e.g., words or word sequences) receive a small, non-zero probability. This avoids assigning zero probability to valid but unseen combinations of words, improving the generalization of the model.

---

### Why is Smoothing Necessary?

Language models estimate the probability of words or sequences of words from a corpus. Without smoothing, any word or sequence not seen in the training data will have a probability of **zero**. This can cause problems such as:

1. **Zero Probability for Valid Events:**
   - For unseen words or bigrams/trigrams, the language model cannot generate or predict them.
   - For example, if "machine learning" never appears together in the training data, a bigram model will assign $P(\text{"learning"}|\text{"machine"}) = 0$, which is unrealistic.

2. **Poor Generalization:**
   - Language models should generalize beyond the training data, but without smoothing, they rely too heavily on observed frequencies and fail to account for rare events.

---

### How Does Smoothing Work?

Smoothing redistributes some probability mass from frequent (or observed) events to rare (or unobserved) events. This ensures that:
- No event has zero probability.
- The overall probability distribution remains valid (i.e., sums to 1).

---

### Common Smoothing Techniques

#### 1. Laplace Smoothing (Add-One Smoothing)

- Adds a constant (usually 1) to the count of each word or n-gram to ensure no probability is zero.
- Formula for unigram:
  $$
  P(w) = \frac{\text{Count}(w) + 1}{|C| + |V|}
  $$
  Where:
  - $|C|$: Total count of all words in the corpus.
  - $|V|$: Vocabulary size.

- **Advantages:**
  - Simple and effective for small vocabularies.
- **Disadvantages:**
  - Overestimates probabilities for rare or unseen events.

---

#### 2. Additive (Add-$\alpha$) Smoothing

- A generalization of Laplace smoothing where a smaller constant $\alpha$ (e.g., 0.1) is added to each count:
  $$
  P(w) = \frac{\text{Count}(w) + \alpha}{|C| + \alpha |V|}
  $$

---

#### 3. Jelinek-Mercer Smoothing (Interpolation)

- Combines higher-order and lower-order models (e.g., bigram + unigram) with a weighted interpolation:
  $$
  P(w_i|w_{i-1}) = \lambda P_{\text{MLE}}(w_i|w_{i-1}) + (1 - \lambda) P(w_i)
  $$
  Where $\lambda$ is a weight parameter.

- **Advantages:**
  - Balances observed data and a fallback model (e.g., unigram probabilities).
- **Disadvantages:**
  - Requires tuning of $\lambda$.

---

#### 4. Dirichlet Prior Smoothing

- Assumes a Dirichlet prior over the word probabilities and smooths them using a background (corpus-wide) model:
  $$
  P(w|D) = \frac{\text{Count}(w, D) + \mu P(w|C)}{|D| + \mu}
  $$
  Where:
  - $\mu$: Smoothing parameter that controls the influence of the background model.
  - $P(w|C)$: Probability of $w$ in the entire collection.

---

#### 5. Kneser-Ney Smoothing

- Redistributes probability mass for higher-order n-grams to lower-order ones, focusing on the diversity of word contexts:
  $$
  P_{\text{KN}}(w_i|w_{i-1}) = \max(\text{Count}(w_{i-1}, w_i) - d, 0) + \lambda P(w_i)
  $$
  Where $d$ is a discounting constant.

- **Advantages:**
  - State-of-the-art for n-gram models.
- **Disadvantages:**
  - Computationally intensive.

---

### Key Principles of Smoothing

1. **Probability Redistribution:**
   - Reduce probabilities of frequent events slightly to assign non-zero probabilities to unseen events.
2. **Balance Between Observed and Unseen Data:**
   - Combine observed counts with general background knowledge of language.
3. **Valid Probability Distribution:**
   - Ensure that the total probability across all events remains equal to 1.

---

### Applications of Smoothing

- **Language Modeling:** For tasks like speech recognition, machine translation, and text generation.
- **Information Retrieval:** In models like Query Likelihood and Divergence-based retrieval.
- **Text Classification:** Where unseen n-grams in test data can appear frequently.

















# Week 3: Information Retrieval Evaluation: 

What you can't measure you can't improve.

Most information retrieval systems are tuned to optimise for an objective evaluation metric. 

**Measures of IR System**; 

- *How fast does it index?* Num documents\hour. Average document size 
- *How fast does it search?* Latency as a function of index size
- *Expressiveness of query language:*  Ability to express complex information needs. Speed on complex queries.

Two different approaches to evaluation, **Online and Offline**: 

**Online**: “Online evaluation is evaluation of a fully functioning system based on implicit
measurement of real users’ experiences of the system in a natural usage environment"

**Offline**: Offline metrics are measured in an isolated environment before deploying a new IR
system. These look at whether a particular set of relevant results are returned when
retrieving items with the system.

|            | Online                                      | Offline                                      |
|------------|---------------------------------------------|----------------------------------------------|
| **Pros**   | Cheap                                       | Fast                                         |
|            | Measure actual user reactions              | Easy to try new ideas                       |
|            |                                             | Amortized Cost Portable                     |
| **Cons**   | Need to go live                             | Needs ground truth                          |
|            | Noisy                                       | Slow at the beginning                       |
|            | Slow                                        | "Expensive"                                 |
|            | Not duplicable                             | Inconsistent                                |
|            |                                             | Difficult to model how users behave         |

- Relevant metrics to determine how good the IR system is at retrieving relevant documents include; 

**Precision**: The fraction of retrieved documents that are relevant = tp / (tp + fp) 
**Recall**: The fraction of relevant documents that are retrieved = tp / (tp + fn)
**Accuracy**: fraction of documents that are relevant and retrieved or not-relevant and not retrieved divided by total number of documents = (tp + fn) / (tp + fp + fn + tn). This is not always useful, especially when there is a class imbalance between relevant documents and non-relevant documents. Accuracy can still be very high while your IR system is sub-optimal. Better to look at precision and recall instead. 
**F-Measure**:

# **Mean Average Precision (MAP) in Information Retrieval**

## **1. Understanding MAP**
**Mean Average Precision (MAP)** is a common evaluation metric for **ranking-based information retrieval systems** (e.g., search engines, recommendation systems). It measures the quality of ranked retrieval results by evaluating **both precision and ranking order**.

### **Key Components:**
- **Precision @ k**: The proportion of relevant documents retrieved in the top-𝑘 results.
- **Average Precision (AP)**: The mean of precision values computed at the ranks where relevant documents appear.
- **Mean Average Precision (MAP)**: The mean of AP scores across multiple queries.

---

## **2. Formula for MAP**
Given a set of **queries** $ Q $, we compute **MAP** as:

$$
MAP = \frac{1}{|Q|} \sum_{q=1}^{|Q|} AP(q)
$$

where:

$$
AP(q) = \frac{1}{R} \sum_{k=1}^{N} P(k) \cdot rel(k)
$$

- $ R $ = Total number of relevant documents for query $ q $
- $ P(k) $ = Precision at position $ k $
- $ rel(k) $ = Binary relevance (1 if the document is relevant, 0 otherwise)
- $ N $ = Total retrieved documents

---

## **3. Example Calculation**
### **Scenario**
A system retrieves **5 documents** for a query, and the **ground truth** has 3 relevant documents.

### **Ranked Retrieval Results:**
| Rank | Document | Relevant? (1=Yes, 0=No) | Precision @ k |
|------|---------|-------------------------|--------------|
| 1    | D1      | ✅ (1)                    | 1.0          |
| 2    | D2      | ❌ (0)                    | 0.5          |
| 3    | D3      | ✅ (1)                    | 0.67         |
| 4    | D4      | ✅ (1)                    | 0.75         |
| 5    | D5      | ❌ (0)                    | 0.6          |

### **Computing Average Precision (AP)**:
\[
AP = \frac{(1.0 + 0.67 + 0.75)}{3} = \frac{2.42}{3} = 0.807
\]

For multiple queries, we compute **AP** for each and take their mean to get **MAP**.

---

## **4. MAP vs Other Metrics**
| Metric  | Description |
|---------|------------|
| **Precision@k** | Measures relevant results in top \( k \), but ignores ranking order. |
| **Recall** | Measures retrieved relevant results but doesn't consider ranking. |
| **F1-score** | Balances precision and recall. |
| **MAP** | Evaluates ranking by averaging precision across relevant documents. |

MAP is **better** than simple **Precision** or **Recall** because it considers the **order of retrieved results**, favoring systems that return **relevant documents earlier**.

---

## **5. When to Use MAP**
✅ **Best suited for:**  
- Search engines, recommendation systems, and ranking problems.  
- Scenarios where **retrieving all relevant documents is important**.  
- Evaluating **ranking quality** over multiple queries.

❌ **Not ideal when:**  
- You only care about the **top-1 or top-k** results (use **Precision@k** instead).  
- A binary decision (relevant/not relevant) is not enough, and graded relevance (e.g., **NDCG**) is required.

---






 # Week 5: Intro to ML and Data Mining
 
 
 
**Data Mining:** The process of discovering (mining) useful patterns from or conducting inferences based on various types of data sources such as structured information repositories (e.g. databases), text, images, sound, video and so on... 

Distinction between ML and data mining becoming increasingly difficult. 

**Association Rule Mining:**

In [22]:
# Mock dataset: Transactions stored in a dictionary
transactions = {
    1: ["Bread", "Milk", "Butter"],
    2: ["Bread", "Butter"],
    3: ["Milk", "Diaper", "Beer"],
    4: ["Bread", "Milk", "Diaper", "Butter"],
    5: ["Milk", "Diaper", "Beer", "Butter"],
    6: ["Bread", "Milk", "Beer"]
}


#Finding Support for each item.
#Support(X) = Transactions containing X / Total Transactions 

def build_support(dic, words):

    support_dic = {}
    
    total_transactions = len(dic)
    
    count = []
    
    counter = 0
    for word in words: 
        num = 0
        for vals in dic.values(): 
            for val in vals: 
                if val == word: 
                    num += 1
                else: 
                    pass
        count.append(num)
        
        support_dic[word] = count[counter] / total_transactions
        counter += 1
         
            
    return support_dic


unique_words = set()
for word in transactions.values():
    unique_words.update(word)


words = list(unique_words)
print(build_support(transactions, words = words))

   

{'Bread': 0.6666666666666666, 'Milk': 0.8333333333333334, 'Diaper': 0.5, 'Beer': 0.5, 'Butter': 0.6666666666666666}


### The next step in the Apriori Algorithm is to generate frequent itemsets and prune infrequent ones.

In [112]:
support_dic = build_support(transactions, words = words)

min_support = 0.5 # must define minimum support: 

frequent_items = {}

for key, value in support_dic.items():
    if value >= min_support: 
        frequent_items[key] = value
    else:
        pass
    
  
print("frequent_items:", frequent_items)
# Now want to generate 

#Generate candidate itemsets of size k from frequent (k-1)-itemsets

def generate_candidates(frequent_itemsets, k):
    """Generate candidate itemsets of size k from frequent (k-1)-itemsets."""
    itemsets = list(frequent_itemsets.keys())  # Convert dictionary keys to a list
    candidates = []  # Store candidate itemsets

    for i in range(len(itemsets)):
        for j in range(i + 1, len(itemsets)):  
            itemset1 = itemsets[i]
            itemset2 = itemsets[j]


            if k == 2:  # Directly form 2-itemsets
                candidates.append((itemset1, itemset2))
            else:  
                # Check if first (k-1) elements are the same (prefix matching)
                if itemset1[:-1] == itemset2[:-1]:  
                    candidate = tuple(sorted(set(itemset1) | set(itemset2)))  # Merge and sort

                    if len(candidate) == k and candidate not in candidates:
                        candidates.append(candidate)

    return candidates

# Generate candidate 2-itemsets
L2_candidates = generate_candidates(frequent_items, k=2)
print("L2 Candidates:", L2_candidates)

# Generate candidate 3-itemsets using L2
L3_candidates = generate_candidates({itemset: 0 for itemset in L2_candidates}, k=3)
print("L3 Candidates:", L3_candidates)
    


frequent_items: {'Bread': 0.6666666666666666, 'Milk': 0.8333333333333334, 'Diaper': 0.5, 'Beer': 0.5, 'Butter': 0.6666666666666666}
L2 Candidates: [('Bread', 'Milk'), ('Bread', 'Diaper'), ('Bread', 'Beer'), ('Bread', 'Butter'), ('Milk', 'Diaper'), ('Milk', 'Beer'), ('Milk', 'Butter'), ('Diaper', 'Beer'), ('Diaper', 'Butter'), ('Beer', 'Butter')]
L3 Candidates: [('Bread', 'Diaper', 'Milk'), ('Beer', 'Bread', 'Milk'), ('Bread', 'Butter', 'Milk'), ('Beer', 'Bread', 'Diaper'), ('Bread', 'Butter', 'Diaper'), ('Beer', 'Bread', 'Butter'), ('Beer', 'Diaper', 'Milk'), ('Butter', 'Diaper', 'Milk'), ('Beer', 'Butter', 'Milk'), ('Beer', 'Butter', 'Diaper')]


In [113]:
transactions = {
    1: ["Bread", "Milk", "Butter"],
    2: ["Bread", "Butter"],
    3: ["Milk", "Diaper", "Beer"],
    4: ["Bread", "Milk", "Diaper", "Butter"],
    5: ["Milk", "Diaper", "Beer", "Butter"],
    6: ["Bread", "Milk", "Beer"]
}


candidates = generate_candidates(frequent_items, k=2)
#print(candidates)

def count_support(transactions, candidates):
    
    support_counts = {}
    for itemset in candidates: 
        support_counts[itemset] = 0
        
    for transaction in transactions.values(): 
        transaction_set = set(transaction)
        for candidate in candidates:
            if set(candidate).issubset(transaction_set): 
                support_counts[candidate] += 1
    
    total_transactions = len(transactions)
    support_values = {}
    
    for candidate in candidates: 
        support_values[candidate] = support_counts[candidate]/total_transactions
    
    return support_values
   

print(count_support(transactions = transactions, candidates = candidates))
        

{('Bread', 'Milk'): 0.5, ('Bread', 'Diaper'): 0.16666666666666666, ('Bread', 'Beer'): 0.16666666666666666, ('Bread', 'Butter'): 0.5, ('Milk', 'Diaper'): 0.5, ('Milk', 'Beer'): 0.5, ('Milk', 'Butter'): 0.5, ('Diaper', 'Beer'): 0.3333333333333333, ('Diaper', 'Butter'): 0.3333333333333333, ('Beer', 'Butter'): 0.16666666666666666}


In [114]:
L2_candidates = generate_candidates(frequent_items, k=2)
print("L2 Candidates:", L2_candidates)
print()
L2_support = count_support(transactions, L2_candidates)
print("L2 Support:", L2_support)

L2 Candidates: [('Bread', 'Milk'), ('Bread', 'Diaper'), ('Bread', 'Beer'), ('Bread', 'Butter'), ('Milk', 'Diaper'), ('Milk', 'Beer'), ('Milk', 'Butter'), ('Diaper', 'Beer'), ('Diaper', 'Butter'), ('Beer', 'Butter')]

L2 Support: {('Bread', 'Milk'): 0.5, ('Bread', 'Diaper'): 0.16666666666666666, ('Bread', 'Beer'): 0.16666666666666666, ('Bread', 'Butter'): 0.5, ('Milk', 'Diaper'): 0.5, ('Milk', 'Beer'): 0.5, ('Milk', 'Butter'): 0.5, ('Diaper', 'Beer'): 0.3333333333333333, ('Diaper', 'Butter'): 0.3333333333333333, ('Beer', 'Butter'): 0.16666666666666666}


In [115]:
def prune_candidates(support_values, min_support): 
    
    pruned_support_values = {}
    
    for candidate, support in support_values.items(): 
        if support >= min_support: 
            pruned_support_values[candidate] = support
        else:
            pass
        
    return pruned_support_values

print(L2_support)
print()

L2_frequent = prune_candidates(L2_support, min_support)
print("Frequent L2 Itemsets:", L2_frequent)            
        
    
    

{('Bread', 'Milk'): 0.5, ('Bread', 'Diaper'): 0.16666666666666666, ('Bread', 'Beer'): 0.16666666666666666, ('Bread', 'Butter'): 0.5, ('Milk', 'Diaper'): 0.5, ('Milk', 'Beer'): 0.5, ('Milk', 'Butter'): 0.5, ('Diaper', 'Beer'): 0.3333333333333333, ('Diaper', 'Butter'): 0.3333333333333333, ('Beer', 'Butter'): 0.16666666666666666}

Frequent L2 Itemsets: {('Bread', 'Milk'): 0.5, ('Bread', 'Butter'): 0.5, ('Milk', 'Diaper'): 0.5, ('Milk', 'Beer'): 0.5, ('Milk', 'Butter'): 0.5}


In [116]:
L3_candidates = generate_candidates(L2_frequent, k=3)
L3_support = count_support(transactions, L3_candidates)
L3_frequent = prune_candidates(L3_support, min_support)
print("Frequent L3 Itemsets:", L3_frequent)

Frequent L3 Itemsets: {}


## Step 4️⃣: Generate Association Rules

Once we have frequent itemsets, we generate association rules of the form:

A $\rightarrow$ B

where:

	•	Confidence measures how often B appears when A is present.
	•	Lift compares confidence to expected frequency.


### **🔹 1️⃣ Confidence Formula**
Confidence measures the **likelihood** of seeing $B$ given that $A$ has occurred.

$$
\text{Confidence}(A \rightarrow B) = \frac{\text{Support}(A \cup B)}{\text{Support}(A)}
$$

**Interpretation**:  
- $\text{Confidence}(A \rightarrow B) = 80\%$ means that when $A$ appears, $B$ also appears **80% of the time** in the dataset.
- Higher confidence suggests a **stronger** association.

✅ **Example Calculation**:  
- $\text{Support}(\{Milk, Bread\}) = 0.4$  
- $\text{Support}(\{Milk\}) = 0.5$  
- $\text{Confidence}(\{Milk\} \rightarrow \{Bread\})$:

$$
\frac{0.4}{0.5} = 0.8 \quad (80\%)
$$

---

### **🔹 2️⃣ Lift Formula**
Lift measures how much **more likely** $B$ appears **when A is present**, compared to $B$ appearing randomly.

$$
\text{Lift}(A \rightarrow B) = \frac{\text{Confidence}(A \rightarrow B)}{\text{Support}(B)}
$$

Or equivalently:

$$
\text{Lift}(A \rightarrow B) = \frac{\text{Support}(A \cup B)}{\text{Support}(A) \times \text{Support}(B)}
$$

**Interpretation**:
- **Lift = 1** → $A$ and $B$ are **independent** (no association).
- **Lift > 1** → $A$ and $B$ occur **more often together than expected** (positive association).
- **Lift < 1** → $A$ actually **reduces** the likelihood of $B$ occurring (negative association).

✅ **Example Calculation**:  
- $\text{Confidence}(\{Milk\} \rightarrow \{Bread\}) = 0.8$
- $\text{Support}(\{Bread\}) = 0.6$
- $\text{Lift}(\{Milk\} \rightarrow \{Bread\})$:

$$
\frac{0.8}{0.6} = 1.33
$$

This means that **buying Milk increases the likelihood of buying Bread by 1.33 times compared to random chance**.

In [122]:
from itertools import combinations
def generate_rules(frequent_itemsets, support_values, L1_support, min_confidence):
    """Generates strong association rules from frequent itemsets."""
    rules = []

    for itemset in frequent_itemsets.keys():
        length = len(itemset)
        if length < 2:
            continue  # Skip single items

        print(f"\nProcessing Itemset: {itemset}")  # Debug print

        # Generate all possible subsets of the itemset
        for i in range(1, length):
            subsets = list(combinations(itemset, i))

            for antecedent in subsets:
                consequent = tuple(sorted(set(itemset) - set(antecedent)))  # Find remaining items
                if not consequent:  # Skip empty consequents
                    continue

                # Use L1_support for single-item antecedents
                antecedent_support = support_values.get(antecedent, L1_support.get(antecedent, None))

                # Debug: Print support values
                print(f"Antecedent: {antecedent}, Consequent: {consequent}")
                print(f"Support(A ∪ B) = {support_values.get(itemset, 'Missing')}, Support(A) = {antecedent_support}")

                if antecedent_support:
                    confidence = support_values[itemset] / antecedent_support
                    print(f"Computed Confidence = {confidence}")

                    if confidence >= min_confidence:
                        rules.append((antecedent, consequent, confidence))

    return rules

min_confidence = 0  # Define minimum confidence threshold
L1_support = count_support(transactions, [(item,) for item in frequent_items.keys()])

association_rules = generate_rules(L2_frequent, L2_support, L1_support, min_confidence)
print("Association Rules:", association_rules)



Processing Itemset: ('Bread', 'Milk')
Antecedent: ('Bread',), Consequent: ('Milk',)
Support(A ∪ B) = 0.5, Support(A) = 0.6666666666666666
Computed Confidence = 0.75
Antecedent: ('Milk',), Consequent: ('Bread',)
Support(A ∪ B) = 0.5, Support(A) = 0.8333333333333334
Computed Confidence = 0.6

Processing Itemset: ('Bread', 'Butter')
Antecedent: ('Bread',), Consequent: ('Butter',)
Support(A ∪ B) = 0.5, Support(A) = 0.6666666666666666
Computed Confidence = 0.75
Antecedent: ('Butter',), Consequent: ('Bread',)
Support(A ∪ B) = 0.5, Support(A) = 0.6666666666666666
Computed Confidence = 0.75

Processing Itemset: ('Milk', 'Diaper')
Antecedent: ('Milk',), Consequent: ('Diaper',)
Support(A ∪ B) = 0.5, Support(A) = 0.8333333333333334
Computed Confidence = 0.6
Antecedent: ('Diaper',), Consequent: ('Milk',)
Support(A ∪ B) = 0.5, Support(A) = 0.5
Computed Confidence = 1.0

Processing Itemset: ('Milk', 'Beer')
Antecedent: ('Milk',), Consequent: ('Beer',)
Support(A ∪ B) = 0.5, Support(A) = 0.8333333333

### **Apriori Algorithm: Association Rule Mining**

The **Apriori algorithm** is used to identify **frequent itemsets** and generate **association rules** in transaction datasets. It follows these steps:

---

## **Step 1️⃣: Compute Support for Single Items (L1)**

- Support of an itemset $X$ is defined as:

$$
\text{Support}(X) = \frac{\text{Transactions containing } X}{\text{Total Transactions}}
$$

- Identify **frequent 1-itemsets** that satisfy the **minimum support threshold** $\text{min\_support}$.

---

## **Step 2️⃣: Generate Candidate Itemsets (L2, L3, …)**

- Using **frequent $(k-1)$-itemsets**, generate **$k$-item candidates**.
- The **Apriori property** states that any **subset** of a frequent itemset **must also be frequent**.

### **Candidate Generation:**
For itemsets of size $k$:
- If two itemsets of size $(k-1)$ share the first $(k-2)$ items, merge them into a $k$-itemset.
- Ensure the new candidate contains **no duplicate elements**.

---

## **Step 3️⃣: Prune Candidates Below `min_support`**
- For each candidate itemset, count its occurrence in transactions.
- Remove candidates that do not satisfy:

$$
\text{Support}(X) \geq \text{min\_support}
$$

- The remaining itemsets are **frequent itemsets**.

---

## **Step 4️⃣: Repeat for Larger Itemsets (`L3`, `L4`, …)**
- Use the **frequent L2 itemsets** to generate **L3 candidates**.
- Compute **support** and **prune**.
- Repeat until **no new frequent itemsets can be formed**.

---

## **Step 5️⃣: Generate Association Rules**
Once frequent itemsets are identified, generate **association rules**:

$$
A \rightarrow B
$$

where:
- **$A$ (Antecedent):** Left-hand side of the rule.
- **$B$ (Consequent):** Right-hand side of the rule.

Rules are evaluated using:

### **1️⃣ Confidence**
Confidence measures the probability of observing $B$ given $A$:

$$
\text{Confidence}(A \rightarrow B) = \frac{\text{Support}(A \cup B)}{\text{Support}(A)}
$$

A high confidence value means that when $A$ appears, $B$ is also likely to appear.

---

### **2️⃣ Lift**
Lift measures how much **more likely** $B$ appears **when A is present**, compared to $B$ appearing randomly.

$$
\text{Lift}(A \rightarrow B) = \frac{\text{Support}(A \cup B)}{\text{Support}(A) \times \text{Support}(B)}
$$

#### **Interpretation:**
- **Lift = 1** → $A$ and $B$ are **independent** (no association).
- **Lift > 1** → $A$ and $B$ occur **more often together than expected** (**positive association**).
- **Lift < 1** → $A$ actually **reduces** the likelihood of $B$ occurring (**negative association**).

---

## **📌 Summary of the Pipeline**
| **Step**  | **What It Does** |
|------------|----------------|
| **Step 1** | Compute **L1 (single-item) support** and filter based on `min_support` |
| **Step 2** | Generate **L2 candidates** using the **Apriori property** |
| **Step 3** | Compute **support for L2** and **prune infrequent itemsets** |
| **Step 4** | **Repeat for L3, L4, ...** until no more candidates exist |
| **Step 5** | Generate **association rules** and compute **confidence & lift** |

---


# Coursework Notes: 


# **Task 1: Evaluating Retrieval Quality (20 marks)**  

This task involves implementing **two ranking metrics**:  
1. **Average Precision (AP)**  
2. **Normalized Discounted Cumulative Gain (NDCG)**  

You will compute these metrics for **BM25 rankings** on the validation set.

---

## **1. Understanding the Metrics**
### **1.1. Average Precision (AP)**
- Measures the quality of ranked retrieval results.
- Formula:

$$
AP = \frac{1}{R} \sum_{k=1}^{N} P(k) \cdot rel(k)
$$

where:
- \( R \) = total number of **relevant** passages.
- \( P(k) \) = precision at rank \( k \).
- \( rel(k) \) = 1 if the passage at rank \( k \) is relevant, else 0.

---

### **1.2. Normalized Discounted Cumulative Gain (NDCG)**
- Measures how well **highly relevant** documents are ranked at the top.
- Formula:

$$
DCG@k = \sum_{i=1}^{k} \frac{rel_i}{\log_2(i+1)}
$$

$$
NDCG@k = \frac{DCG@k}{IDCG@k}
$$

where:
- \( rel_i \) = relevance score at rank \( i \).
- \( IDCG@k \) = ideal (best possible) DCG at **cutoff \( k \)**.

---

## **2. Implementing the Metrics**
Below is the Python implementation:

```python

import numpy as np

def average_precision(relevance_list):
    """
    Compute Average Precision (AP).
    
    relevance_list: List of binary relevance scores (1 for relevant, 0 for not relevant)
    
    Returns:
        AP score
    """
    num_relevant = sum(relevance_list)
    if num_relevant == 0:
        return 0.0

    precision_sum = 0.0
    relevant_count = 0

    for i, rel in enumerate(relevance_list):
        if rel == 1:
            relevant_count += 1
            precision_at_i = relevant_count / (i + 1)  # Precision at rank i
            precision_sum += precision_at_i

    return precision_sum / num_relevant

def dcg_at_k(relevance_list, k):
    """
    Compute Discounted Cumulative Gain (DCG) at rank k.
    
    relevance_list: List of relevance scores (binary or graded)
    k: Rank cutoff
    
    Returns:
        DCG@k score
    """
    relevance_list = np.array(relevance_list[:k])
    discounts = np.log2(np.arange(2, len(relevance_list) + 2))
    return np.sum(relevance_list / discounts)

def ndcg_at_k(relevance_list, k):
    """
    Compute Normalized Discounted Cumulative Gain (NDCG) at rank k.
    
    relevance_list: List of relevance scores (binary or graded)
    k: Rank cutoff
    
    Returns:
        NDCG@k score
    """
    dcg_k = dcg_at_k(relevance_list, k)
    ideal_relevance_list = sorted(relevance_list, reverse=True)  # Ideal ranking
    idcg_k = dcg_at_k(ideal_relevance_list, k)
    
    return dcg_k / idcg_k if idcg_k > 0 else 0.0

# Example usage with dummy relevance list
relevance_example = [1, 0, 1, 1, 0, 0, 1]  # Example relevance scores

ap_score = average_precision(relevance_example)
ndcg_score = ndcg_at_k(relevance_example, k=5)

ap_score, ndcg_score

```

# **Numerical Example for Average Precision (AP) and NDCG**

Let's go step by step with a small **ranked list of passages** and their **relevance scores**.

## **Example Setup**
Suppose we have **5 retrieved passages** for a query, and their **binary relevance scores** (1 for relevant, 0 for non-relevant) are:

| Rank | Passage ID | Relevance ($rel$) |
|------|-----------|----------------|
| 1    | P1        | 1              |
| 2    | P2        | 0              |
| 3    | P3        | 1              |
| 4    | P4        | 1              |
| 5    | P5        | 0              |

Now, let's compute **AP and NDCG@5** step by step.

---

## **1. Calculating Average Precision (AP)**
The formula for Average Precision is:

$$
AP = \frac{1}{R} \sum_{k=1}^{N} P(k) \cdot rel(k)
$$

where:
- $R$ = total number of relevant passages = **3** (P1, P3, P4).
- $P(k)$ = precision at rank $k$.
- $rel(k)$ = 1 if passage at rank $k$ is relevant, else 0.

### **Step-by-step calculation**
Compute **precision at each rank** where $rel(k) = 1$:

| Rank k | Relevance rel(k) | Precision P(k) | Contribution to AP |
|------|------------|----------------|-----------------------------|
| 1    | 1          | $ 1/1 = 1.0 $  | $ 1.0 \times 1 = 1.0 $ |
| 2    | 0          | -              | - |
| 3    | 1          | $ 2/3 = 0.667 $| $ 0.667 \times 1 = 0.667 $ |
| 4    | 1          | $ 3/4 = 0.75 $ | $ 0.75 \times 1 = 0.75 $ |
| 5    | 0          | -              | - |

Now, compute the final **AP**:

$$
AP = \frac{(1.0 + 0.667 + 0.75)}{3} = \frac{2.417}{3} = 0.806
$$

So, **AP = 0.806**.

---

## **2. Calculating NDCG@5**
The formula for **Discounted Cumulative Gain (DCG)** is:

$$
DCG@k = \sum_{i=1}^{k} \frac{rel_i}{\log_2(i+1)}
$$

And **Normalized DCG (NDCG)** is:

$$
NDCG@k = \frac{DCG@k}{IDCG@k}
$$

where $IDCG@k$ is the **ideal** DCG (i.e., the best possible ranking).

### **Step-by-step calculation**
Compute **DCG@5**:

$$
DCG@5 = \frac{1}{\log_2(1+1)} + \frac{0}{\log_2(2+1)} + \frac{1}{\log_2(3+1)} + \frac{1}{\log_2(4+1)} + \frac{0}{\log_2(5+1)}
$$

$$
= \frac{1}{\log_2(2)} + \frac{0}{\log_2(3)} + \frac{1}{\log_2(4)} + \frac{1}{\log_2(5)} + \frac{0}{\log_2(6)}
$$

$$
= \frac{1}{1} + \frac{0}{1.585} + \frac{1}{2} + \frac{1}{2.322} + \frac{0}{2.585}
$$

$$
= 1 + 0 + 0.5 + 0.43 + 0 = 1.93
$$

### **Computing Ideal DCG (IDCG@5)**
The **ideal** ranking is the one with all relevant passages ranked first:

| Rank | Ideal Relevance |
|------|----------------|
| 1    | 1              |
| 2    | 1              |
| 3    | 1              |
| 4    | 0              |
| 5    | 0              |

$$
IDCG@5 = \frac{1}{\log_2(1+1)} + \frac{1}{\log_2(2+1)} + \frac{1}{\log_2(3+1)} + \frac{0}{\log_2(4+1)} + \frac{0}{\log_2(5+1)}
$$

$$
= \frac{1}{1} + \frac{1}{1.585} + \frac{1}{2} + \frac{0}{2.322} + \frac{0}{2.585}
$$

$$
= 1 + 0.63 + 0.5 + 0 + 0 = 2.13
$$

### **Final NDCG@5 Calculation**
$$
NDCG@5 = \frac{DCG@5}{IDCG@5} = \frac{1.93}{2.13} = 0.906
$$

So, **NDCG@5 = 0.906**.

---

## **Final Results**
For this example:
- **AP = 0.806**
- **NDCG@5 = 0.906**

---



# Task 2: Logistic Regression

## Overview: 

**Key Steps**
1) 	Represent queries and passages using word embeddings
    
            •	Use Word2Vec, GloVe, FastText, or ELMo to obtain embeddings for words.
            •	Compute query and passage embeddings by averaging word embeddings.
            
2)	Implement Logistic Regression (from scratch)

            •	Define a logistic regression function to predict passage relevance.
            •	Train using gradient descent.
            •	Evaluate using AP and NDCG.
            
3) Analyze Learning Rate

            •	Train with different learning rates.
            •	Observe impact on training loss over epochs.
            
4)	Consider Negative Sampling (Optional)
	
            •	If training data is imbalanced, generate additional negative examples.

# **Numerical Example: Computing Query & Passage Embeddings**
Let's go step by step with a small numerical example to **manually compute query and passage embeddings**.

---

## **Step 1: Define a Small Vocabulary with Word Embeddings**
Assume we have a small **word embedding dictionary** where each word is represented by a **3-dimensional vector** (instead of 300d for simplicity).

| Word        | Embedding (3D) |
|------------|---------------------|
| machine    | **[0.2, 0.4, 0.6]**  |
| learning   | **[0.3, 0.7, 0.5]**  |
| deep       | **[0.1, 0.3, 0.9]**  |
| subset     | **[0.5, 0.2, 0.1]**  |
| techniques | **[0.6, 0.8, 0.4]**  |

---

## **Step 2: Tokenize the Query and Passage**
We have:

**Query:** `"machine learning techniques"`  
**Passage:** `"deep learning is a subset of machine learning"`

After tokenization and removing stopwords:

Query Tokens = [“machine”, “learning”, “techniques”]
Passage Tokens = [“deep”, “learning”, “subset”, “machine”, “learning”]


---

## **Step 3: Retrieve Word Embeddings**
Using the table above, we retrieve the embeddings for each word.

### **Query Embeddings:**
- `machine` → **[0.2, 0.4, 0.6]**
- `learning` → **[0.3, 0.7, 0.5]**
- `techniques` → **[0.6, 0.8, 0.4]**

### **Passage Embeddings:**
- `deep` → **[0.1, 0.3, 0.9]**
- `learning` → **[0.3, 0.7, 0.5]**
- `subset` → **[0.5, 0.2, 0.1]**
- `machine` → **[0.2, 0.4, 0.6]**
- `learning` → **[0.3, 0.7, 0.5]** (again)

---

## **Step 4: Compute Mean Embeddings**
We compute the **mean vector** for the query and passage by averaging the embeddings.

### **Query Embedding:**

$$
E_q = \frac{1}{3} \sum_{i=1}^{3} v_i
$$

$$
E_q = \frac{1}{3} \left( [0.2, 0.4, 0.6] + [0.3, 0.7, 0.5] + [0.6, 0.8, 0.4] \right)
$$

$$
= \frac{1}{3} [ (0.2+0.3+0.6), (0.4+0.7+0.8), (0.6+0.5+0.4) ]
$$

$$
= \frac{1}{3} [ 1.1, 1.9, 1.5 ]
$$

$$
= [0.367, 0.633, 0.5]
$$

---

### **Passage Embedding:**

$$
E_p = \frac{1}{5} \sum_{i=1}^{5} v_i
$$

$$
E_p = \frac{1}{5} \left( [0.1, 0.3, 0.9] + [0.3, 0.7, 0.5] + [0.5, 0.2, 0.1] + [0.2, 0.4, 0.6] + [0.3, 0.7, 0.5] \right)
$$

$$
= \frac{1}{5} [ (0.1+0.3+0.5+0.2+0.3), (0.3+0.7+0.2+0.4+0.7), (0.9+0.5+0.1+0.6+0.5) ]
$$

$$
= \frac{1}{5} [ 1.4, 2.3, 2.6 ]
$$

$$
= [0.28, 0.46, 0.52]
$$

---

## **Final Output: Embeddings**
- **Query Embedding:** **[0.367, 0.633, 0.5]**
- **Passage Embedding:** **[0.28, 0.46, 0.52]**

These **fixed-length embeddings** now represent the query and passage numerically and will be used as input for **Logistic Regression**.

---

# **Logistic Regression for Passage Relevance**

Now that we have **query and passage embeddings**, we can implement **Logistic Regression** to predict whether a passage is **relevant** to a given query.

---

## **1. Logistic Regression Overview**
Logistic Regression is a binary classification model that estimates the probability that a passage is relevant to a query.

### **Mathematical Formulation**
Given an input feature vector $ x $, the **logistic regression model** predicts a probability $ \hat{y} $:

$$
\hat{y} = \sigma(w^T x + b)
$$

where:
- $ x $ = **[query embedding ⊕ passage embedding]** (concatenation of query & passage vectors)
- $ w $ = **trainable weight vector** (to be learned)
- $ b $ = **bias term** (trainable scalar)
- $ \sigma(z) = \frac{1}{1 + e^{-z}} $ = **sigmoid function**

### **Loss Function: Binary Cross-Entropy**
Since this is a binary classification task (relevant = 1, not relevant = 0), we optimize the **binary cross-entropy loss**:

$$
\mathcal{L} = - \frac{1}{N} \sum_{i=1}^{N} y_i \log(\hat{y}_i) + (1 - y_i) \log(1 - \hat{y}_i)
$$

where:
- $ y_i $ = ground truth (1 if relevant, 0 otherwise)
- $ \hat{y}_i $ = predicted probability

---

## **2. Feature Representation**
Each input sample consists of:
1. **Query embedding** $ E_q $ (300D)
2. **Passage embedding** $ E_p $ (300D)
3. **Final input vector**:

$$
x = [E_q \oplus E_p]
$$

This results in a **600-dimensional input vector** ($ x \in \mathbb{R}^{600} $).

---

## **3. Training Logistic Regression**
### **Step 1: Initialize Parameters**
We initialize the weight vector $ w $ and bias $ b $:

$$
w \in \mathbb{R}^{600}, \quad b \in \mathbb{R}
$$

### **Step 2: Forward Pass**
For each query-passage pair, compute the probability of relevance:

$$
z = w^T x + b
$$

$$
\hat{y} = \sigma(z) = \frac{1}{1 + e^{-z}}
$$

### **Step 3: Compute Loss**
The **binary cross-entropy loss** measures how well the model's predicted probabilities match the ground truth:

$$
\mathcal{L} = - \frac{1}{N} \sum_{i=1}^{N} y_i \log(\hat{y}_i) + (1 - y_i) \log(1 - \hat{y}_i)
$$

### **Step 4: Compute Gradients**
We update the parameters using **gradient descent**:

- **Gradient of the loss with respect to weights:**

$$
\frac{\partial \mathcal{L}}{\partial w} = \frac{1}{N} \sum_{i=1}^{N} x_i (\hat{y}_i - y_i)
$$

- **Gradient of the loss with respect to bias:**

$$
\frac{\partial \mathcal{L}}{\partial b} = \frac{1}{N} \sum_{i=1}^{N} (\hat{y}_i - y_i)
$$

### **Step 5: Update Weights**
Using **gradient descent**, update $ w $ and $ b $:

$$
w = w - \eta \frac{\partial \mathcal{L}}{\partial w}
$$

$$
b = b - \eta \frac{\partial \mathcal{L}}{\partial b}
$$

where $ \eta $ is the **learning rate**.

---

## **4. Evaluating the Model**
Once the model is trained, we evaluate it using **ranking metrics**.

### **Step 1: Compute Predictions**
For each query-passage pair, we obtain a probability $ \hat{y} $ and predict relevance:

$$
\hat{y} = \sigma(w^T x + b)
$$

If $ \hat{y} \geq 0.5 $, classify as **relevant (1)**, otherwise **not relevant (0)**.

### **Step 2: Compute AP and NDCG**
We use **Average Precision (AP) and NDCG** to measure ranking performance.

- **Average Precision (AP):**

$$
AP = \frac{1}{R} \sum_{k=1}^{N} P(k) \cdot rel(k)
$$

where:
- $ P(k) $ = Precision at rank $ k $
- $ rel(k) $ = 1 if the passage at rank $ k $ is relevant

- **NDCG@k:**

$$
DCG@k = \sum_{i=1}^{k} \frac{rel_i}{\log_2(i+1)}
$$

$$
NDCG@k = \frac{DCG@k}{IDCG@k}
$$

where $ IDCG@k $ is the ideal ranking.

---

## **5. Analyzing Learning Rate**
The learning rate $ \eta $ controls how much we update $ w $ and $ b $ at each step.

### **Effects of Learning Rate:**
1. **Too small ($ \eta \to 0 $)** → Slow convergence
2. **Too large ($ \eta \to 1 $)** → Instability, oscillations
3. **Optimal ($ \eta \approx 0.01 $)** → Smooth loss decrease

To analyze this, we plot **loss vs. epochs** for different learning rates:

- $ \eta = 0.001 $
- $ \eta = 0.01 $
- $ \eta = 0.1 $

---

## **6. Summary**
✔ **Implemented Logistic Regression theory**  
✔ **Defined feature representation using embeddings**  
✔ **Derived gradient descent updates**  
✔ **Explained evaluation using AP & NDCG**  
✔ **Analyzed learning rate effects**

---

# **Numerical Example: Logistic Regression for Passage Relevance**
Let’s go step by step through a **manual numerical example** to illustrate how logistic regression predicts passage relevance.

---

## **Step 1: Define a Small Feature Space**
Instead of using **600-dimensional embeddings**, we use a simple **3D feature space** for easier calculations.

Suppose we have the following **query and passage embeddings**:

| Feature | Query Embedding ($E_q$) | Passage Embedding ($E_p$) |
|---------|----------------|----------------|
| $x_1$  | **0.2**        | **0.4**        |
| $x_2$  | **0.3**        | **0.7**        |
| $x_3$  | **0.5**        | **0.6**        |

The **feature vector $x$** is formed by concatenating query and passage embeddings:

$$
x = [E_q \oplus E_p] = [0.2, 0.3, 0.5, 0.4, 0.7, 0.6]
$$

---

## **Step 2: Initialize Logistic Regression Parameters**
We initialize **weights $w$** and **bias $b$**:

$$
w = [0.1, 0.2, -0.1, 0.05, -0.3, 0.4]
$$

$$
b = 0.05
$$

---

## **Step 3: Compute Weighted Sum ($z$)**
The model computes:

$$
z = w^T x + b
$$

$$
= (0.1 \times 0.2) + (0.2 \times 0.3) + (-0.1 \times 0.5) + (0.05 \times 0.4) + (-0.3 \times 0.7) + (0.4 \times 0.6) + 0.05
$$

$$
= (0.02) + (0.06) + (-0.05) + (0.02) + (-0.21) + (0.24) + 0.05
$$

$$
= -0.07
$$

---

## **Step 4: Apply the Sigmoid Function**
We compute the probability:

$$
\hat{y} = \sigma(z) = \frac{1}{1 + e^{-z}}
$$

$$
= \frac{1}{1 + e^{0.07}}
$$

Approximating:

$$
e^{0.07} \approx 1.0725
$$

$$
\hat{y} = \frac{1}{1 + 1.0725} = \frac{1}{2.0725} \approx 0.483
$$

---

## **Step 5: Predict Passage Relevance**
We use a **threshold of 0.5**:
- If $ \hat{y} \geq 0.5 $, classify as **relevant (1)**.
- If $ \hat{y} < 0.5 $, classify as **not relevant (0)**.

Since $ \hat{y} = 0.483 $, the model **predicts that the passage is NOT relevant (0).**

---

## **Step 6: Compute the Loss (Binary Cross-Entropy)**
Assume the **true relevance label** is $ y = 1 $ (meaning the passage is actually relevant).

The **Binary Cross-Entropy Loss** is:

$$
\mathcal{L} = - y \log(\hat{y}) - (1 - y) \log(1 - \hat{y})
$$

$$
= - (1 \times \log(0.483)) - (0 \times \log(1 - 0.483))
$$

$$
= - \log(0.483)
$$

Approximating:

$$
\log(0.483) \approx -0.315
$$

$$
\mathcal{L} = 0.315
$$

---

## **Step 7: Compute Gradients for Weight Updates**
To update $ w $ and $ b $, we compute the gradients:

$$
\frac{\partial \mathcal{L}}{\partial w} = x (\hat{y} - y)
$$

$$
\frac{\partial \mathcal{L}}{\partial b} = (\hat{y} - y)
$$

Since:

$$
\hat{y} - y = 0.483 - 1 = -0.517
$$

We compute the weight updates:

$$
\Delta w = -0.517 \times x = [-0.103, -0.155, -0.259, -0.207, -0.362, -0.31]
$$

Bias update:

$$
\Delta b = -0.517
$$

---

## **Step 8: Update Weights (Gradient Descent)**
Using **learning rate** $ \eta = 0.01 $:

$$
w = w - \eta \Delta w
$$

$$
b = b - \eta \Delta b
$$

$$
w = [0.1, 0.2, -0.1, 0.05, -0.3, 0.4] - 0.01 \times [-0.103, -0.155, -0.259, -0.207, -0.362, -0.31]
$$

$$
= [0.10103, 0.20155, -0.09741, 0.05207, -0.29638, 0.4031]
$$

$$
b = 0.05 - 0.01 \times (-0.517) = 0.050517
$$

---

## **Final Summary**
1. **Initial prediction:** $ \hat{y} = 0.483 \Rightarrow $ Model predicts **not relevant (0)**.
2. **True label:** $ y = 1 $ (actual passage is relevant).
3. **Loss:** $ 0.315 $ (cross-entropy loss).
4. **Weight updates:** Weights and bias are **slightly updated** after **gradient descent**.

---

# **LambdaMART: Learning to Rank with XGBoost**

## **1. What is LambdaMART?**
LambdaMART is a **gradient-boosted tree-based ranking algorithm**, combining:
- **LambdaRank** (a ranking optimization method using pairwise relevance differences)
- **MART (Multiple Additive Regression Trees)** (a boosting-based tree ensemble model)

LambdaMART is **superior to Logistic Regression** because:
✔ It captures **non-linear relationships** between features.  
✔ It optimizes **ranking metrics** directly (NDCG, MAP).  
✔ It can learn from **graded relevance labels** (not just binary labels).  

---

## **2. Input Representation for LambdaMART**
To train the **LambdaMART ranking model**, we need:

1. **Feature Vectors:**  
   - **Query features**: Query embedding (300D)
   - **Passage features**: Passage embedding (300D)
   - **BM25 score** (optional)
   - **Query-Passage similarity features** (e.g., cosine similarity)

2. **Relevance Scores:**  
   - Instead of **binary labels (0/1)** from Logistic Regression, we can use **graded relevance scores** (if available).
   - If only **binary labels** exist, we can still train the model effectively.

Final **feature vector (X) format**:
$$
X = [E_q \oplus E_p, \text{BM25 score}, \text{Cosine Similarity}]
$$
Where:
- $ E_q $ → Query embedding (300D)
- $ E_p $ → Passage embedding (300D)
- **BM25 Score (1D)**
- **Cosine Similarity (1D)**  

Thus, **input size = 600 + 2 = 602 dimensions**.

---

## **3. Training LambdaMART with XGBoost**
XGBoost allows **LambdaMART training** by setting:

$$
\text{objective} = "rank:ndcg"
$$

This tells XGBoost to **optimize NDCG (Normalized Discounted Cumulative Gain)** directly.

### **Training Setup**
- **Query Groups**:  
  - XGBoost requires queries to be grouped so it **knows which passages belong to which query**.
- **Hyperparameter Tuning**:  
  - Important parameters include:
    - `learning_rate`: Step size in gradient boosting.
    - `n_estimators`: Number of boosting rounds.
    - `max_depth`: Tree depth.
    - `min_child_weight`: Minimum samples required to split a node.

---

## **4. LambdaMART Model Training Pipeline**
### **Step 1: Preprocessing**
- Compute **query and passage embeddings** using **Word2Vec, GloVe, or FastText**.
- Compute **BM25 scores** (if available).
- Compute **Cosine Similarity**:

$$
\text{CosineSim}(E_q, E_p) = \frac{E_q \cdot E_p}{||E_q|| \cdot ||E_p||}
$$

- Form feature matrix **$X$** (size **$N \times 602$**) and corresponding **relevance labels**.

### **Step 2: Define Query Groups**
- XGBoost requires knowing how many **passages belong to each query**.
- Example: If **Query 1 has 3 passages** and **Query 2 has 2 passages**, the query group is:

$$
\text{query_groups} = [3, 2]
$$

### **Step 3: Train LambdaMART**
- Set $ \text{objective}="rank:ndcg" $
- Tune **hyperparameters** (learning rate, max depth, etc.).

---

## **5. Hyperparameter Tuning Methodology**
We optimize the following hyperparameters:

| Hyperparameter | Description |
|---------------|-------------|
| `learning_rate` | Step size for boosting (e.g., 0.01, 0.05, 0.1) |
| `n_estimators` | Number of boosting rounds (e.g., 100, 200, 500) |
| `max_depth` | Maximum depth of trees (e.g., 3, 5, 7) |
| `min_child_weight` | Minimum samples required to split a node |
| `subsample` | Fraction of data used per boosting round (0.7-1.0) |

### **Grid Search Strategy**
1. **Step 1:** Train models with different `learning_rate` values: **(0.01, 0.05, 0.1)**
2. **Step 2:** For the best `learning_rate`, tune `max_depth`: **(3, 5, 7)**
3. **Step 3:** Optimize `n_estimators`: **(100, 200, 500)**
4. **Step 4:** Fine-tune `min_child_weight` and `subsample`

We evaluate each model using **NDCG@10** on the **validation set**.

---

# **6. Numerical Example**
Let's assume we have **one query with three passages**, each represented by three features.

| Query ID | Passage ID | Feature 1 ($x_1$) | Feature 2 ($x_2$) | Feature 3 ($x_3$) | Relevance ($y$) |
|----------|-----------|----------------|----------------|----------------|----------------|
| 1        | 101       | **0.2**        | **0.5**        | **0.7**        | **2** |
| 1        | 102       | **0.1**        | **0.4**        | **0.6**        | **1** |
| 1        | 103       | **0.3**        | **0.2**        | **0.8**        | **0** |

Here, **higher $y$ values indicate higher relevance**.

### **Step 1: Compute Pairwise Differences**
LambdaMART optimizes ranking by computing **gradient updates** for **pairwise differences**:

- **Compare (101, 102)** → **101 should be ranked higher**.
- **Compare (101, 103)** → **101 should be ranked higher**.
- **Compare (102, 103)** → **102 should be ranked higher**.

### **Step 2: Compute NDCG**
Assume the model initially ranks the passages as:

| Rank | Passage ID | Model Score |
|------|-----------|--------------|
| 1    | 103       | **0.9** |
| 2    | 102       | **0.7** |
| 3    | 101       | **0.5** |

The **DCG (Discounted Cumulative Gain)** is:

$$
DCG = \sum_{i=1}^{k} \frac{rel_i}{\log_2(i+1)}
$$

$$
DCG = \frac{0}{\log_2(1+1)} + \frac{1}{\log_2(2+1)} + \frac{2}{\log_2(3+1)}
$$

$$
= 0 + \frac{1}{1.585} + \frac{2}{2}
$$

$$
= 0 + 0.63 + 1 = 1.63
$$

The **ideal ranking (IDCG)** would have **passage 101 at the top**, so:

$$
NDCG = \frac{DCG}{IDCG}
$$

---

# **7. Final Summary**
✔ **Implemented LambdaMART using XGBoost**  
✔ **Defined input representation (query embeddings, passage embeddings, BM25, cosine similarity)**  
✔ **Explained hyperparameter tuning strategy**  
✔ **Demonstrated a simple numerical example with NDCG calculation**

---




# **Neural Network Model for Passage Re-Ranking with Binary Relevance Scores**

## **1. Problem Definition**
We train a **neural network model** to **re-rank passages** based on their **binary relevance scores** (0 or 1).  
Unlike LambdaMART, which uses **boosted trees**, we now use **deep learning architectures** such as:
- **Feedforward Neural Networks (MLP)**
- **CNNs**
- **RNNs (LSTM/GRU)**
- **Transformers (BERT-based models)**

For efficiency, we use **Feedforward MLP** first.

---

## **2. Input Representation**
Each query-passage pair is represented as:

$$
X = [E_q \oplus E_p, \text{BM25 score}, \text{Cosine Similarity}]
$$

where:
- **$E_q$** → Query embedding (300D)
- **$E_p$** → Passage embedding (300D)
- **BM25 Score (1D)**
- **Cosine Similarity (1D)**  

Total **input size**: **602 dimensions**.

The **label ($Y$)** is **binary relevance**:  
- **1 = relevant**
- **0 = not relevant**

---

## **3. Choosing the Neural Network Architecture**
We choose a **Feedforward Neural Network (MLP)** with:
1. **Input Layer (602D)**
2. **Hidden Layers (Dense Fully Connected Layers)**
   - **ReLU activations** for non-linearity.
   - **Dropout** for regularization.
3. **Output Layer (Sigmoid Activation)**
   - Outputs a **probability score** (0 to 1).

---

## **4. Mathematical Formulation**
Given input **$X$**, the neural network computes:

$$
h_1 = \text{ReLU}(W_1 X + b_1)
$$

$$
h_2 = \text{ReLU}(W_2 h_1 + b_2)
$$

$$
\hat{y} = \sigma(W_3 h_2 + b_3)
$$

where:
- $ W_1, W_2, W_3 $ = Trainable weight matrices
- $ b_1, b_2, b_3 $ = Bias terms
- $ \sigma(z) $ = **Sigmoid function** to get a probability score.

---

## **5. Loss Function: Binary Cross-Entropy**
Since **relevance is binary (0 or 1)**, we use **Binary Cross-Entropy Loss**:

$$
\mathcal{L} = - \frac{1}{N} \sum_{i=1}^{N} y_i \log(\hat{y}_i) + (1 - y_i) \log(1 - \hat{y}_i)
$$

where:
- $ y_i \in \{0,1\} $ is the **ground truth** (1 = relevant, 0 = not relevant).
- $ \hat{y}_i $ is the **predicted probability of relevance**.

The goal is to:
✔ **Maximize** probability for **relevant passages ($y = 1$)**.  
✔ **Minimize** probability for **non-relevant passages ($y = 0$)**.

---

## **6. Evaluation Metrics**
After training, evaluate model performance using **ranking metrics**:
1. **NDCG@k** (Ranking Quality)
2. **MAP (Mean Average Precision)**

To compute **NDCG**, sort passages **by predicted probability** and calculate:

$$
DCG@k = \sum_{i=1}^{k} \frac{rel_i}{\log_2(i+1)}
$$

$$
NDCG@k = \frac{DCG@k}{IDCG@k}
$$

---

# **7. Numerical Example**
### **Step 1: Example Query-Passage Pairs**
Assume we have **one query with three passages**, each represented by three features:

| Query ID | Passage ID | Feature 1 ($x_1$) | Feature 2 ($x_2$) | Feature 3 ($x_3$) | Relevance ($y$) |
|----------|-----------|----------------|----------------|----------------|----------------|
| 1        | 101       | **0.2**        | **0.5**        | **0.7**        | **1** |
| 1        | 102       | **0.1**        | **0.4**        | **0.6**        | **1** |
| 1        | 103       | **0.3**        | **0.2**        | **0.8**        | **0** |

### **Step 2: Compute Model Predictions**
The neural network assigns probability scores:

| Rank | Passage ID | Model Score ($\hat{y}$) |
|------|-----------|--------------|
| 1    | 103       | **0.9** |
| 2    | 102       | **0.7** |
| 3    | 101       | **0.4** |

However, this ranking is **incorrect**, as passages **101 and 102** are **more relevant** than **103**.

### **Step 3: Compute NDCG**
We calculate **DCG** based on predicted ranking:

$$
DCG = \sum_{i=1}^{k} \frac{rel_i}{\log_2(i+1)}
$$

$$
DCG = \frac{0}{\log_2(1+1)} + \frac{1}{\log_2(2+1)} + \frac{1}{\log_2(3+1)}
$$

$$
= 0 + \frac{1}{1.585} + \frac{1}{2}
$$

$$
= 0 + 0.63 + 0.5 = 1.13
$$

The **ideal ranking (IDCG)** would have **passages 101 and 102 ranked first**, giving:

$$
IDCG = 1 + \frac{1}{\log_2(2+1)} + \frac{0}{\log_2(3+1)}
$$

$$
= 1 + \frac{1}{1.585} + 0 = 1.63
$$

Thus, **NDCG is**:

$$
NDCG = \frac{DCG}{IDCG} = \frac{1.13}{1.63} = 0.69
$$

The model’s ranking is **not optimal** (NDCG is **0.69 instead of 1**), indicating **incorrect ordering**.

---

## **8. Final Summary**
✔ **Designed a Feedforward MLP for passage ranking**  
✔ **Defined loss function (Binary Cross-Entropy) for binary labels**  
✔ **Explained feature representation (query + passage embeddings, BM25, cosine similarity)**  
✔ **Provided a numerical example with ranking computation using NDCG**  
✔ **Compared predictions with the ideal ranking**  

---
